### PR TITLE
[#13858][fix] AutoDeploy fix the piecewise vlm issue

### PR DESCRIPTION
**Pull request by nvchenghaoz**
#13858 

Enable the piecewise CG by default to see if there are any further errors. 

The main fix is to move the Gamma related op to the dynamic split. Also update the kwargs capture to only capture the static related ones.. 

## Summary by CodeRabbit

* **New Features**
  * Added support for Gemma4 multimodal mask preparation in CUDA graph compilation.

* **Improvements**
  * Optimized buffer allocation for captured graphs to allocate memory only for model parameters actually consumed by compiled submodules, reducing memory overhead.

* **Configuration Changes**
  * Enabled piecewise CUDA graph compilation by default.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/14006)

### CODE REVIEW

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1,DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1"

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1"

**Comment by suyoggupta** [comment]
maybe we can try adding a smoke test https://sourcegraph.com/r/github.com/NVIDIA/TensorRT-LLM/-/blob/tests/unittest/auto_deploy/singlegpu/smoke/test_ad_build_small_single.py ?

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1"

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

File: tests/integration/defs/accuracy/test_llm_api_autodeploy.py [hunk 1]
```diff
@@ -35,6 +35,22 @@
...
+        return
+    config.setdefault("transforms",
+                      {}).setdefault("compile_model",
+                                     {})["piecewise_enabled"] = False
+
+
+def _disable_piecewise(config: dict) -> None:
+    config.setdefault("transforms",
+                      {}).setdefault("compile_model",
+                                     {})["piecewise_enabled"] = False
```
Comment (nvchenghaoz) [inline tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: These are not needed given the changes in the `llm_args.py`

File: tensorrt_llm/_torch/auto_deploy/llm_args.py [hunk 2]
```diff
@@ -63,6 +63,7 @@ def _check_for_default_value_only(
     "attn_backend": ("insert_cached_attention.backend", "transformers_replace_cached_attn.backend"),
     "compile_backend": ("compile_model.backend",),
 }
+_PIECEWISE_COMPILE_BACKENDS = {"torch-cudagraph", "torch-opt"}
```
Comment (nvchenghaoz) [inline tensorrt_llm/_torch/auto_deploy/llm_args.py]: Move this to the function

File: tests/unittest/auto_deploy/_utils_test/_model_test_utils.py [hunk 3]
```diff
@@ -611,6 +611,20 @@ def apply_rotary_pos_emb_ds(q, k, cos, sin, position_ids, unsqueeze_dim=1):
     },
 }
 
+_PIECEWISE_COMPILE_BACKENDS = {"torch-cudagraph", "torch-opt"}
```
Comment (nvchenghaoz) [inline tests/unittest/auto_deploy/_utils_test/_model_test_utils.py]: same here, these are not needed right? or we should have a unified place for these util functions

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

File: tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py [hunk 4]
```diff
@@ -32,127 +33,144 @@
...
+        The dynamic op still needs to be a split boundary, but piecewise should
+        not add output stabilization. This covers inplace ops that return None,
+        ops that already return persistent/stable buffers, logits or host-driven
+        helpers whose outputs are not consumed by captured static segments, and
+        request-dependent outputs whose shapes are not safely bounded by the
+        piecewise token bucket.
+    """
+
+    # Wrap with DynamicOpWrapper and pass a stable out= buffer.
+    OUT_BUFFER = auto()
```
Comment (suyoggupta) [inline tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py:83]: Have we checked how much additional buffer space we need?

File: tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py [hunk 5]
```diff
@@ -32,127 +33,144 @@
...
+_DYNAMIC_OP_POLICIES: Dict[str, DynamicOpPolicy] = {
+    # Dynamic kernels whose outputs feed captured static segments.
+    "auto_deploy::flashinfer_attention_mha_with_cache": DynamicOpPolicy.OUT_BUFFER,
+    "auto_deploy::triton_attention_flattened_mha_with_cache": DynamicOpPolicy.OUT_BUFFER,
+    "auto_deploy::triton_paged_mha_with_cache": DynamicOpPolicy.OUT_BUFFER,
+    "auto_deploy::torch_cached_attention_with_cache": DynamicOpPolicy.OUT_BUFFER,
+    "auto_deploy::trtllm_attention_mha_with_cache": DynamicOpPolicy.OUT_BUFFER,
+    "auto_deploy::flashinfer_mla_with_cache": DynamicOpPolicy.OUT_BUFFER,
+    "auto_deploy::flashinfer_trtllm_mla_with_cache": DynamicOpPolicy.OUT_BUFFER,
+    "auto_deploy::torch_cached_mla_with_cache": DynamicOpPolicy.OUT_BUFFER,
```
Comment (suyoggupta) [inline tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py:100]: maintaining this decoupled from the op defn feels like a footgun and a persistent source of issues. can we think about annotating the custom op directly with `DynamicOpPolicy` info?

File: tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py (hunk 4 above)
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py:83]: This part of the change is mainly the refactoring. 

OUT_BUFFER is part of the graph pool and will be re-used. 

The METADATA_WRAPPER is the place that may increase some memory usage. But since the METADATA is small, the extra mem usage is pretty small.

File: tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py (hunk 5 above)
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py:100]: If we add new operators, the effort to support piecewise CG for new operators for either approaches should be the same as they may require some understanding of the actual handling for different Policy. And require some extra code updates. I can move this to custom op if you prefer this way.

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Review by jieli-matrix — APPROVED** [review]

File: tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py (hunk 5 above)
Comment (suyoggupta) [reply tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py:100]: can we add a todo to revisit this decision? Doing it the way it's done here keeps the op-defn clean and not tied to a particular PCG impl.

**Review by suyoggupta — APPROVED** [review]

File: tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py (hunk 5 above)
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/compile/piecewise_utils.py:100]: done.

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast

**Comment by nvchenghaoz** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1" --disable-fail-fast
